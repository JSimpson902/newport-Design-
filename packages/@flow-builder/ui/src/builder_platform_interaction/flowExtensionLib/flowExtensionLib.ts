// @ts-nocheck
export * from './flowExtension';
