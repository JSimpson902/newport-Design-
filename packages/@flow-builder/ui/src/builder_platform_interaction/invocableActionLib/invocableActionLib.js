import {
    fetchOnce,
    SERVER_ACTION_TYPE
} from 'builder_platform_interaction/serverDataLib';
import {
    FLOW_AUTOMATIC_OUTPUT_HANDLING,
    getProcessTypeAutomaticOutPutHandlingSupport
} from 'builder_platform_interaction/processTypeLib';

let cachedDetails = [];
let invocableActions = [];

export function setInvocableActions(actions) {
    invocableActions = actions;
}

export function getInvocableActions() {
    return invocableActions;
}

export function fetchDetailsForInvocableAction(
    { actionName, actionType },
    { background = false, disableErrorModal = false, messageForErrorModal } = {}
) {
    const key = `${actionName}-${actionType}`;
    if (cachedDetails[key]) {
        return Promise.resolve(cachedDetails[key]);
    }
    const params = { actionName, actionType };
    return fetchOnce(
        SERVER_ACTION_TYPE.GET_INVOCABLE_ACTION_DETAILS,
        params,
        {
            background,
            disableErrorModal,
            messageForErrorModal
        }
    ).then(details => {
        cachedDetails[key] = details;
        return details;
    });
}

/**
 * Grabs the parameters for a specific invocable action from the cache
 *
 * @param {Object} actionId - the action identifier
 * @param {string} actionId.actionName - the action name
 * @param {string} actionId.actionType - the action type
 * @returns {Object} the action parameters
 */
export function getParametersForInvocableAction({ actionName, actionType }) {
    const key = `${actionName}-${actionType}`;
    return cachedDetails[key] && cachedDetails[key].parameters;
}

export function isAutomaticOutputHandlingSupported(flowProcessType) {
    const processTypeAutomaticOutPutHandlingSupport = getProcessTypeAutomaticOutPutHandlingSupport(
        flowProcessType
    );
    return (
        processTypeAutomaticOutPutHandlingSupport ===
        FLOW_AUTOMATIC_OUTPUT_HANDLING.SUPPORTED
    );
}

export function clearInvocableActionCachedParameters() {
    cachedDetails = [];
}
