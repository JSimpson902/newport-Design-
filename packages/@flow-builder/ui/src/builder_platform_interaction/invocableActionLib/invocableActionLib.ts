// @ts-nocheck
import { getValueFromHydratedItem } from 'builder_platform_interaction/dataMutationLib';
import { ParameterListRowItem } from 'builder_platform_interaction/elementFactory';
import {
    FLOW_AUTOMATIC_OUTPUT_HANDLING,
    getProcessTypeAutomaticOutPutHandlingSupport
} from 'builder_platform_interaction/processTypeLib';
import { fetchOnce, SERVER_ACTION_TYPE } from 'builder_platform_interaction/serverDataLib';

let cachedDetails = [];
let standardInvocableActions = [];
let dynamicInvocableActions = [];

export type InvocableAction = {
    elementType: ELEMENT_TYPE;
    actionType: string | ValueWithError;
    actionName: string | ValueWithError;
    inputParameters: ParameterListRowItem[];
    outputParameters: ParameterListRowItem[];
    displayText?: string | ValueWithError;
    name?: string | ValueWithError;
    label?: string | ValueWithError;
};
/**
 * Set standard invocable actions.
 *
 * @param actions - The standard actions to set
 */
export function setStandardInvocableActions(actions) {
    standardInvocableActions = actions;
}

/**
 * Set dynamic invocable actions.
 *
 * @param actions - The dynamic actions to set
 */
export function setDynamicInvocableActions(actions) {
    dynamicInvocableActions = actions;
}

/**
 * @returns standard invocable actions
 */
export function getStandardInvocableActions() {
    return standardInvocableActions;
}

/**
 * @returns dynamic invocable actions
 */
export function getDynamicInvocableActions() {
    return dynamicInvocableActions;
}

/**
 * @returns invocable actions
 */
export function getInvocableActions(): InvocableAction[] {
    return standardInvocableActions.concat(dynamicInvocableActions);
}

/**
 * @returns invocable actions and corresponding custom icon resource
 */
export function getActionIconMap() {
    return standardInvocableActions
        .concat(dynamicInvocableActions)
        .filter((action) => action.iconName)
        .reduce((acc, action) => {
            acc[action.name] = action.iconName;
            return acc;
        }, {});
}

/**
 * @param root0
 * @param root0.actionName
 * @param root0.actionType
 * @param root1
 * @param root1.background
 * @param root1.disableErrorModal
 * @param root1.messageForErrorModal
 */
export function fetchDetailsForInvocableAction(
    { actionName, actionType },
    { background = false, disableErrorModal = false, messageForErrorModal = undefined } = {}
) {
    const key = getActionKey(actionName, actionType);

    if (cachedDetails[key]) {
        return Promise.resolve(cachedDetails[key]);
    }
    const params = { actionName, actionType };
    return fetchOnce(SERVER_ACTION_TYPE.GET_INVOCABLE_ACTION_DETAILS, params, {
        background,
        disableErrorModal,
        messageForErrorModal
    }).then((details) => {
        cachedDetails[key] = details;
        return details;
    });
}

/**
 * Get the key for an action
 *
 * @param actionName - The action name
 * @param actionType - The action type
 * @returns  the action key, or undefined if not a valid action
 */
export function getActionKey(actionName?: string, actionType?: string): string | undefined {
    if (actionName == null || actionType == null) {
        return undefined;
    }

    return `${actionName}-${actionType}`;
}

/**
 * Updates a data type and sets a subtype in generically-typed extension parameters
 *
 * @param {*} parameters - Screen field parameters
 * @param {*} dynamicTypeMappings - A mapping of generic types to concrete types
 */
export function applyDynamicTypeMappings(parameters, dynamicTypeMappings) {
    if (!dynamicTypeMappings || dynamicTypeMappings.length === 0) {
        return parameters;
    }

    return parameters
        .map((parameter) => ({
            parameter,
            dynamicTypeMapping: dynamicTypeMappings.find(
                (dynamicTypeMapping) => parameter.sobjectType === getValueFromHydratedItem(dynamicTypeMapping.typeName)
            )
        }))
        .map(({ parameter, dynamicTypeMapping }) => {
            if (dynamicTypeMapping && dynamicTypeMapping.typeValue) {
                parameter = Object.assign({}, parameter, {
                    sobjectType: getValueFromHydratedItem(dynamicTypeMapping.typeValue)
                });
            }
            return parameter;
        });
}

/**
 * Grabs the parameters for a specific invocable action from the cache
 *
 * @param {Object} actionId - the action identifier
 * @param {string} actionId.actionName - the action name
 * @param {string} actionId.actionType - the action type
 * @param actionId.dataTypeMappings
 * @returns {Object} the action parameters
 */
export function getParametersForInvocableAction({ actionName, actionType, dataTypeMappings }) {
    const key = getActionKey(actionName, actionType);
    const params = cachedDetails[key] && cachedDetails[key].parameters;
    return applyDynamicTypeMappings(params, dataTypeMappings);
}

/**
 * @param flowProcessType
 */
export function isAutomaticOutputHandlingSupported(flowProcessType) {
    const processTypeAutomaticOutPutHandlingSupport = getProcessTypeAutomaticOutPutHandlingSupport(flowProcessType);
    return processTypeAutomaticOutPutHandlingSupport === FLOW_AUTOMATIC_OUTPUT_HANDLING.SUPPORTED;
}

/**
 *
 */
export function clearInvocableActionCachedParameters() {
    cachedDetails = [];
}
