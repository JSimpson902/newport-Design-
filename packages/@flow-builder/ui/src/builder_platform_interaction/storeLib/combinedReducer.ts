// @ts-nocheck
/**
 * Combine multiple reducer to a one reducer function.
 * @param {Object} reducers object containing key value pair of all the reducers. Key is name of the property of the state of a store on which reduction needs to happen.
 * @returns {Function} one reducer function.
 */
export function combinedReducer(reducers) {
    return function combined(state = {}, action) {
        let updated = false;
        const newState = {};
        const reducerKeys = Object.keys(reducers);
        reducerKeys.forEach(key => {
            newState[key] = reducers[key](state[key], action);
            updated = updated || newState[key] !== state[key];
        });
        return updated ? newState : state;
    };
}
