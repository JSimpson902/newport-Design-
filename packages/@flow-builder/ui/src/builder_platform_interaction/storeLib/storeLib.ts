// @ts-nocheck
export { combinedReducer } from './combinedReducer';
export { createSelector } from './createSelector';
export { Store } from './store';
export { generateGuid } from './guidGenerator';
export { deepCopy } from './deepCopy';
export { isPlainObject } from './isPlainObject';
