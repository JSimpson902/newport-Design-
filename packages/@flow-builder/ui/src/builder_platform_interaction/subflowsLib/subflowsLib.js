import { getMergedInputOutputVariables } from './subflowVariablesMerger';
import {
    fetchOnce,
    SERVER_ACTION_TYPE
} from 'builder_platform_interaction/serverDataLib';
import { LABELS } from './subflowsLibLabels';
import { MERGE_VARIABLES_WARNING_TYPE } from './subflowVariablesWarnings';
import { format } from 'builder_platform_interaction/commonUtils';

export { MERGE_VARIABLES_WARNING_TYPE } from './subflowVariablesWarnings';
export { getMergedInputOutputVariables } from './subflowVariablesMerger';

let subflows = [];
let cachedInputOutputVariables = {};
let cachedMergedInputOutputVariables = {};

export function setSubflows(data) {
    subflows = data;
}

export function getSubflows() {
    return subflows;
}

/**
 * Fetch raw input and output variables for given flow name
 * @param {String} flowName the flow name
 * @returns {Promise<FlowInputOutputVariablesVersion[]>} promise to raw input output variables, as returned by the GET_FLOW_INPUT_OUTPUT_VARIABLES service
 */
export function fetchFlowInputOutputVariables(
    flowName,
    { background = false, disableErrorModal = false, messageForErrorModal } = {}
) {
    if (cachedInputOutputVariables[flowName]) {
        return Promise.resolve(cachedInputOutputVariables[flowName]);
    }
    const serverActionParams = { flowName };
    return fetchOnce(
        SERVER_ACTION_TYPE.GET_FLOW_INPUT_OUTPUT_VARIABLES,
        serverActionParams,
        {
            background,
            disableErrorModal,
            messageForErrorModal
        }
    ).then(inputOutputVariables => {
        cachedInputOutputVariables[flowName] = inputOutputVariables;
        return inputOutputVariables;
    });
}

/**
 * Fetch input variables and returns merged latest/active input variables
 * @param {String} flowName the flow name
 * @returns {Promise<MergedFlowVariable[]>}
 */
export function fetchMergedFlowInputVariables(
    flowName,
    { background = false, disableErrorModal = false, messageForErrorModal } = {}
) {
    return fetchFlowInputOutputVariables(flowName, {
        background,
        disableErrorModal,
        messageForErrorModal
    }).then(() => getMergedFlowInputVariables(flowName));
}

/**
 * @typedef {MergedFlowVariable}
 * @property {string} apiName
 * @property {string} dataType
 * @property {boolean} isCollection
 * @property {number} scale
 * @property {string} subtype
 * @property {MERGE_VARIABLES_WARNING_TYPE} mergeWarning
 */

/**
 * Fetch output variables and returns merged latest/active output variables
 * @param {String} flowName the flow name
 * @returns {Promise<MergedFlowVariable[]>}
 */
export function fetchMergedFlowOutputVariables(
    flowName,
    { background = false, disableErrorModal = false, messageForErrorModal } = {}
) {
    return fetchFlowInputOutputVariables(flowName, {
        background,
        disableErrorModal,
        messageForErrorModal
    }).then(() => getMergedFlowOutputVariables(flowName));
}

/**
 * Get the raw input output variables
 * @param {String} flowName the flow name
 * @returns {FlowInputOutputVariablesVersion[] | undefined} raw input output variables or undefined if not yet retrieved
 */
export function getFlowInputOutputVariables(flowName) {
    return cachedInputOutputVariables[flowName];
}

/**
 * Get the merged flow input variables
 * @param {String} flowName the flow name
 * @returns {MergedFlowVariable[] | undefined} the merged flow input variables or undefined if not yet retrieved or unknown flow name
 */
export function getMergedFlowInputVariables(flowName) {
    if (cachedMergedInputOutputVariables[flowName]) {
        return cachedMergedInputOutputVariables[flowName].inputVariables;
    }
    const flowInputOutputVariables = getFlowInputOutputVariables(flowName);
    if (!flowInputOutputVariables) {
        return undefined;
    }
    cachedMergedInputOutputVariables[flowName] = getMergedInputOutputVariables(
        flowInputOutputVariables
    );
    return cachedMergedInputOutputVariables[flowName].inputVariables;
}

/**
 * Get the merged flow output variables
 * @param {String} flowName the flow name
 * @returns {MergedFlowVariable[] | undefined} the merged flow output variables or undefined if not yet retrieved or unknown flow name
 */
export function getMergedFlowOutputVariables(flowName) {
    if (cachedMergedInputOutputVariables[flowName]) {
        return cachedMergedInputOutputVariables[flowName].outputVariables;
    }
    const flowInputOutputVariables = getFlowInputOutputVariables(flowName);
    if (!flowInputOutputVariables) {
        return undefined;
    }
    cachedMergedInputOutputVariables[flowName] = getMergedInputOutputVariables(
        flowInputOutputVariables
    );
    return cachedMergedInputOutputVariables[flowName].outputVariables;
}

/**
 * Get label for a flow variable (with warning if any)
 * @param {MergedFlowVariable} variable
 * @returns {string} the label
 */
export function getSubflowVariableLabelWithWarning(variable) {
    if (variable.mergeWarning) {
        switch (variable.mergeWarning) {
            case MERGE_VARIABLES_WARNING_TYPE.ONLY_AVAILABLE_IN_LATEST:
                return format(
                    LABELS.variableInLatestVersionOnly,
                    variable.apiName
                );
            case MERGE_VARIABLES_WARNING_TYPE.ONLY_AVAILABLE_IN_ACTIVE:
            case MERGE_VARIABLES_WARNING_TYPE.DATA_TYPE_CHANGED:
                return format(
                    LABELS.variableInActiveVersionOnly,
                    variable.apiName
                );
            default:
                return variable.apiName;
        }
    }
    return variable.apiName;
}

export function clearFlowCachedInputOutputVariables() {
    cachedInputOutputVariables = {};
    cachedMergedInputOutputVariables = {};
}
