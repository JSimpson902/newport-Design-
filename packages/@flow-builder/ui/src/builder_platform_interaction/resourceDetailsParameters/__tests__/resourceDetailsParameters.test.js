import { createElement } from 'lwc';
import { describeExtension } from 'builder_platform_interaction/flowExtensionLib';
import { fetchMergedFlowOutputVariables } from 'builder_platform_interaction/subflowsLib';
import { fetchDetailsForInvocableAction } from 'builder_platform_interaction/invocableActionLib';
import { mockSubmitForApprovalActionParameters, mockLocalActionParameters } from 'mock/calloutData';
import ResourceDetailsParameters from 'builder_platform_interaction/resourceDetailsParameters';
import {
    mockExtensionScreenfieldAutomaticOutputsModeResourceDetails,
    mockActionSubmitForApprovalAutomaticOutputsModeResourceDetails,
    mockActionLocalActionInAutomaticOutputsModeResourceDetails,
    mockSubflowInAutomaticOutputModeResourceDetails
} from 'mock/resourceDetailsData';
import { mockFlowRuntimeEmailFlowExtensionDescription } from 'mock/flowExtensionsData';
import { Store } from 'builder_platform_interaction/storeLib';
import { flowWithAllElementsUIModel } from 'mock/storeData';
import { flowWithActiveAndLatest as mockFlowWithActiveAndLatest } from 'serverData/GetFlowInputOutputVariables/flowWithActiveAndLatest.json';

const createComponentUnderTest = resourceDetails => {
    const el = createElement('builder_platform_interaction-resource-details-parameters', {
        is: ResourceDetailsParameters
    });
    el.resourceDetails = resourceDetails;
    document.body.appendChild(el);
    return el;
};

jest.mock('builder_platform_interaction/flowExtensionLib', () => ({
    describeExtension: jest.fn(() => Promise.resolve(mockFlowRuntimeEmailFlowExtensionDescription))
}));

jest.mock('builder_platform_interaction/invocableActionLib', () => ({
    fetchDetailsForInvocableAction: jest.fn(() => Promise.resolve(mockSubmitForApprovalActionParameters))
}));

jest.mock('builder_platform_interaction/subflowsLib', () => {
    const actual = require.requireActual('builder_platform_interaction/subflowsLib');
    return {
        fetchMergedFlowOutputVariables: jest.fn(flowName => {
            if (flowName === 'flowWithActiveAndLatest') {
                return Promise.resolve(
                    actual.getMergedInputOutputVariables(mockFlowWithActiveAndLatest).outputVariables
                );
            }
            return Promise.reject(`No flow with name ${flowName}`);
        }),
        getSubflowVariableLabelWithWarning: actual.getSubflowVariableLabelWithWarning
    };
});

jest.mock('builder_platform_interaction/storeLib', () => require('builder_platform_interaction_mocks/storeLib'));

jest.mock(
    '@salesforce/label/FlowBuilderSubflows.variableInLatestVersionOnly',
    () => ({ default: '{0} (Latest Version Only)' }),
    { virtual: true }
);

jest.mock(
    '@salesforce/label/FlowBuilderSubflows.variableInActiveVersionOnly',
    () => ({ default: '{0} (Active Version Only)' }),
    { virtual: true }
);

const SELECTORS = { SPINNER: '.spinner' };

const EXPECTED_MOCK_ORDERED_PARAMETERS_FOR_LC_EMAIL_IN_AUTO_MODE = [
    {
        apiName: 'disabled',
        label: 'Disabled',
        description: 'Prevents the user from modifying or copying the value.',
        typeIconName: 'utility:crossfilter'
    },
    {
        apiName: 'label',
        label: 'Label',
        description: 'The label that appears above the email field.',
        typeIconName: 'utility:text'
    },
    {
        apiName: 'placeholder',
        label: 'Placeholder Text',
        description:
            "Text that appears in the field when it's empty. Use placeholder text to give users a hint about what to enter in the field.",
        typeIconName: 'utility:text'
    },
    {
        apiName: 'readonly',
        label: 'Read Only',
        description: 'Prevents the user from modifying the value, but not from copying it.',
        typeIconName: 'utility:crossfilter'
    },
    {
        apiName: 'required',
        label: 'Required',
        description: 'Requires the user to enter a value.',
        typeIconName: 'utility:crossfilter'
    },
    {
        apiName: 'value',
        label: 'Value',
        description:
            "To provide a default value, set this attribute's value. To use the user-entered value elsewhere in your flow, store this attribute's output value in a variable.",
        typeIconName: 'utility:text'
    }
];

const EXPECTED_MOCK_ORDERED_PARAMETERS_FOR_ACTION_SUBMIT_FOR_APPROVAL_IN_AUTO_MODE = [
    {
        apiName: 'instanceId',
        label: 'Instance ID',
        description: 'The ID of the approval process instance.',
        typeIconName: 'utility:text'
    },
    {
        apiName: 'instanceStatus',
        label: 'Instance Status',
        description:
            'The status of the approval. The valid values are "Approved," "Rejected," "Removed," or "Pending."',
        typeIconName: 'utility:text'
    },
    {
        apiName: 'newWorkItemIds',
        label: 'New Work Item IDs',
        description: 'An array of the ID(s) of the work item(s) created for the next step in this approval process.',
        typeIconName: 'utility:text'
    },
    {
        apiName: 'actorIds',
        label: 'Next Approver IDs',
        description: 'An array of the ID(s) of the next approver(s).',
        typeIconName: 'utility:text'
    },
    {
        apiName: 'entityId',
        label: 'Record ID',
        description: 'The ID of the record submitted for approval.',
        typeIconName: 'utility:text'
    }
];

const EXPECTED_MOCK_ORDERED_PARAMETERS_FOR_ACTION_LOCAL_ACTION_IN_AUTO_MODE = [
    {
        apiName: 'greeting',
        label: 'greeting',
        description: null,
        typeIconName: 'utility:text'
    },
    {
        apiName: 'subject',
        label: 'subject',
        description: null,
        typeIconName: 'utility:text'
    }
];

const EXPECTED_MOCK_ORDERED_PARAMETERS_FOR_SUBFLOW_IN_AUTO_MODE = [
    { apiName: 'accountOutput', label: 'accountOutput', typeIconName: 'utility:sobject' },
    { apiName: 'accountOutputCollection', label: 'accountOutputCollection', typeIconName: 'utility:sobject' },
    { apiName: 'carOutput', label: 'carOutput (Latest Version Only)', typeIconName: 'utility:apex' },
    {
        apiName: 'carOutputCollection',
        label: 'carOutputCollection (Latest Version Only)',
        typeIconName: 'utility:apex'
    },
    { apiName: 'inputOutput1', label: 'inputOutput1', typeIconName: 'utility:text' },
    { apiName: 'inputOutput2', label: 'inputOutput2 (Active Version Only)', typeIconName: 'utility:text' },
    { apiName: 'output1', label: 'output1', typeIconName: 'utility:text' },
    { apiName: 'output2', label: 'output2 (Active Version Only)', typeIconName: 'utility:text' },
    { apiName: 'output3', label: 'output3 (Active Version Only)', typeIconName: 'utility:text' },
    { apiName: 'output4', label: 'output4 (Latest Version Only)', typeIconName: 'utility:text' }
];

describe('Resource Details parameters', () => {
    let resourceDetailsParametersComponent;
    beforeAll(() => {
        Store.setMockState(flowWithAllElementsUIModel);
    });
    afterAll(() => {
        Store.resetStore();
    });
    describe('Extension (ie: ligthning component) screenfield in automatic outputs mode', () => {
        describe('No fetch exception', () => {
            beforeEach(() => {
                resourceDetailsParametersComponent = createComponentUnderTest(
                    mockExtensionScreenfieldAutomaticOutputsModeResourceDetails
                );
            });
            describe('Parameters fetch server call OK and NO error', () => {
                test('check "Parameters" details (via API)', () => {
                    const parameters = resourceDetailsParametersComponent.parameters;
                    expect(parameters).toEqual(EXPECTED_MOCK_ORDERED_PARAMETERS_FOR_LC_EMAIL_IN_AUTO_MODE);
                });
                test('check UI: icon names, tooltip, labels...(snapshot) parameters displayed', () => {
                    expect(resourceDetailsParametersComponent).toMatchSnapshot();
                });
            });
            describe('Parameters fetch server call OK but error', () => {
                beforeAll(() => {
                    describeExtension.mockImplementation(() =>
                        Promise.reject(new Error('An error occured during extension parameters fetching'))
                    );
                });
                test('check "Parameters" details (via API)', () => {
                    const parameters = resourceDetailsParametersComponent.parameters;
                    expect(parameters).toHaveLength(0);
                });
                test('check UI: icon names, tooltip, labels (snapshot) no parameters displayed', () => {
                    expect(resourceDetailsParametersComponent).toMatchSnapshot();
                });
            });
        });
        describe('Fetch exception', () => {
            test('spinner should not be displayed)', () => {
                describeExtension.mockImplementation(() => {
                    throw new Error('Runtime exception this time');
                });
                resourceDetailsParametersComponent = createComponentUnderTest(
                    mockExtensionScreenfieldAutomaticOutputsModeResourceDetails
                );
                const spinner = resourceDetailsParametersComponent.shadowRoot.querySelector(SELECTORS.spinner);
                expect(spinner).toBeNull();
            });
        });
    });

    describe('core action', () => {
        describe('"submit for approval" in automatic outputs mode', () => {
            describe('No fetch exception', () => {
                beforeEach(() => {
                    resourceDetailsParametersComponent = createComponentUnderTest(
                        mockActionSubmitForApprovalAutomaticOutputsModeResourceDetails
                    );
                });
                describe('Parameters fetch server call OK and NO error', () => {
                    test('check "Parameters" details (via API)', () => {
                        const parameters = resourceDetailsParametersComponent.parameters;
                        expect(parameters).toEqual(
                            EXPECTED_MOCK_ORDERED_PARAMETERS_FOR_ACTION_SUBMIT_FOR_APPROVAL_IN_AUTO_MODE
                        );
                    });
                    test('check UI: icon names, tooltip, labels...(snapshot) parameters displayed', () => {
                        expect(resourceDetailsParametersComponent).toMatchSnapshot();
                    });
                });
                describe('Parameters fetch server call OK but error', () => {
                    beforeAll(() => {
                        fetchDetailsForInvocableAction.mockImplementation(() =>
                            Promise.reject(new Error('An error occured during extension parameters fetching'))
                        );
                    });
                    test('check "Parameters" details (via API)', () => {
                        const parameters = resourceDetailsParametersComponent.parameters;
                        expect(parameters).toHaveLength(0);
                    });
                    test('check UI: icon names, tooltip, labels (snapshot) no parameters displayed', () => {
                        expect(resourceDetailsParametersComponent).toMatchSnapshot();
                    });
                });
            });
        });

        describe('"Local action" in automatic outputs mode', () => {
            describe('No fetch exception', () => {
                beforeAll(() => {
                    fetchDetailsForInvocableAction.mockImplementation(() => Promise.resolve(mockLocalActionParameters));
                });
                beforeEach(() => {
                    resourceDetailsParametersComponent = createComponentUnderTest(
                        mockActionLocalActionInAutomaticOutputsModeResourceDetails
                    );
                });
                describe('Parameters fetch server call OK and NO error', () => {
                    test('check "Parameters" details (via API)', () => {
                        const parameters = resourceDetailsParametersComponent.parameters;
                        expect(parameters).toEqual(
                            EXPECTED_MOCK_ORDERED_PARAMETERS_FOR_ACTION_LOCAL_ACTION_IN_AUTO_MODE
                        );
                    });
                    test('check UI: icon names, tooltip, labels...(snapshot) parameters displayed', () => {
                        expect(resourceDetailsParametersComponent).toMatchSnapshot();
                    });
                });
                describe('Parameters fetch server call OK but error', () => {
                    beforeAll(() => {
                        fetchDetailsForInvocableAction.mockImplementation(() =>
                            Promise.reject(new Error('An error occured during extension parameters fetching'))
                        );
                    });
                    test('check "Parameters" details (via API)', () => {
                        const parameters = resourceDetailsParametersComponent.parameters;
                        expect(parameters).toHaveLength(0);
                    });
                    test('check UI: icon names, tooltip, labels (snapshot) no parameters displayed', () => {
                        expect(resourceDetailsParametersComponent).toMatchSnapshot();
                    });
                });
            });
        });
    });

    describe('Subflow in automatic outputs mode', () => {
        describe('No fetch exception', () => {
            beforeEach(() => {
                resourceDetailsParametersComponent = createComponentUnderTest(
                    mockSubflowInAutomaticOutputModeResourceDetails
                );
            });
            describe('Parameters fetch server call OK and NO error', () => {
                test('check "Parameters" details (via API)', () => {
                    const parameters = resourceDetailsParametersComponent.parameters;
                    expect(parameters).toEqual(EXPECTED_MOCK_ORDERED_PARAMETERS_FOR_SUBFLOW_IN_AUTO_MODE);
                });
                test('check UI: icon names, tooltip, labels...(snapshot) parameters displayed', () => {
                    expect(resourceDetailsParametersComponent).toMatchSnapshot();
                });
            });
            describe('Parameters fetch server call OK but error', () => {
                beforeAll(() => {
                    fetchMergedFlowOutputVariables.mockImplementation(() =>
                        Promise.reject(new Error('An error occured during subflow parameters fetching'))
                    );
                });
                test('check "Parameters" details (via API)', () => {
                    const parameters = resourceDetailsParametersComponent.parameters;
                    expect(parameters).toHaveLength(0);
                });
                test('check UI: icon names, tooltip, labels (snapshot) no parameters displayed', () => {
                    expect(resourceDetailsParametersComponent).toMatchSnapshot();
                });
            });
        });
        describe('Fetch exception', () => {
            test('spinner should not be displayed)', () => {
                describeExtension.mockImplementation(() => {
                    throw new Error('Runtime exception this time');
                });
                resourceDetailsParametersComponent = createComponentUnderTest(
                    mockSubflowInAutomaticOutputModeResourceDetails
                );
                const spinner = resourceDetailsParametersComponent.shadowRoot.querySelector(SELECTORS.spinner);
                expect(spinner).toBeNull();
            });
        });
    });
});
