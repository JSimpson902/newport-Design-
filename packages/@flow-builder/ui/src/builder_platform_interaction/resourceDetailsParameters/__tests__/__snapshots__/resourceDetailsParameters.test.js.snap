// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource Details parameters Extension (ie: ligthning component) screenfield in automatic outputs mode No fetch exception Parameters fetch server call OK and NO error check UI: icon names, tooltip, labels...(snapshot) parameters displayed 1`] = `
<builder_platform_interaction-resource-details-parameters>
  #shadow-root(open)
    <div
      class="resource-detail-parameters-body"
    >
      <div
        class="slds-text-title_caps slds-text-color_weak slds-p-vertical_x-small"
      >
        FlowBuilderResourceDetailsParameters.header
      </div>
      <ul
        class="slds-has-block-links test-list-section-items"
      >
        <li
          class="slds-item slds-truncate_container"
        >
          <lightning-icon
            class="slds-m-right_xx-small"
          >
            #shadow-root(open)
              <lightning-primitive-icon>
                #shadow-root(open)
                  <svg
                    data-key="utility:crossfilter"
                  >
                    <use />
                  </svg>
              </lightning-primitive-icon>
          </lightning-icon>
          <span
            class="slds-text-body_regular slds-align-middle slds-truncate"
            title="Prevents the user from modifying or copying the value."
          >
            Disabled
          </span>
        </li>
        <li
          class="slds-item slds-truncate_container"
        >
          <lightning-icon
            class="slds-m-right_xx-small"
          >
            #shadow-root(open)
              <lightning-primitive-icon>
                #shadow-root(open)
                  <svg
                    data-key="utility:text"
                  >
                    <use />
                  </svg>
              </lightning-primitive-icon>
          </lightning-icon>
          <span
            class="slds-text-body_regular slds-align-middle slds-truncate"
            title="The label that appears above the email field."
          >
            Label
          </span>
        </li>
        <li
          class="slds-item slds-truncate_container"
        >
          <lightning-icon
            class="slds-m-right_xx-small"
          >
            #shadow-root(open)
              <lightning-primitive-icon>
                #shadow-root(open)
                  <svg
                    data-key="utility:text"
                  >
                    <use />
                  </svg>
              </lightning-primitive-icon>
          </lightning-icon>
          <span
            class="slds-text-body_regular slds-align-middle slds-truncate"
            title="Text that appears in the field when it's empty. Use placeholder text to give users a hint about what to enter in the field."
          >
            Placeholder Text
          </span>
        </li>
        <li
          class="slds-item slds-truncate_container"
        >
          <lightning-icon
            class="slds-m-right_xx-small"
          >
            #shadow-root(open)
              <lightning-primitive-icon>
                #shadow-root(open)
                  <svg
                    data-key="utility:crossfilter"
                  >
                    <use />
                  </svg>
              </lightning-primitive-icon>
          </lightning-icon>
          <span
            class="slds-text-body_regular slds-align-middle slds-truncate"
            title="Prevents the user from modifying the value, but not from copying it."
          >
            Read Only
          </span>
        </li>
        <li
          class="slds-item slds-truncate_container"
        >
          <lightning-icon
            class="slds-m-right_xx-small"
          >
            #shadow-root(open)
              <lightning-primitive-icon>
                #shadow-root(open)
                  <svg
                    data-key="utility:crossfilter"
                  >
                    <use />
                  </svg>
              </lightning-primitive-icon>
          </lightning-icon>
          <span
            class="slds-text-body_regular slds-align-middle slds-truncate"
            title="Requires the user to enter a value."
          >
            Required
          </span>
        </li>
        <li
          class="slds-item slds-truncate_container"
        >
          <lightning-icon
            class="slds-m-right_xx-small"
          >
            #shadow-root(open)
              <lightning-primitive-icon>
                #shadow-root(open)
                  <svg
                    data-key="utility:text"
                  >
                    <use />
                  </svg>
              </lightning-primitive-icon>
          </lightning-icon>
          <span
            class="slds-text-body_regular slds-align-middle slds-truncate"
            title="To provide a default value, set this attribute's value. To use the user-entered value elsewhere in your flow, store this attribute's output value in a variable."
          >
            Value
          </span>
        </li>
      </ul>
    </div>
</builder_platform_interaction-resource-details-parameters>
`;

exports[`Resource Details parameters Extension (ie: ligthning component) screenfield in automatic outputs mode No fetch exception Parameters fetch server call OK but error check UI: icon names, tooltip, labels (snapshot) no parameters displayed 1`] = `
<builder_platform_interaction-resource-details-parameters>
  #shadow-root(open)
</builder_platform_interaction-resource-details-parameters>
`;
